service/history/statemachines/executor.go

```diff
@@ -0,0 +1,90 @@
+// The MIT License
+//
+// Copyright (c) 2024 Temporal Technologies Inc.  All rights reserved.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package statemachines
+
+import (
+	"context"
+	"time"
+
+	persistencespb "go.temporal.io/server/api/persistence/v1"
+	"go.temporal.io/server/common/definition"
+)
+
+// TaskWithReference is a [Task] with a reference to a statemachine on a specific workflow.
+type TaskWithReference struct {
+	WorkflowKey       definition.WorkflowKeyLike
+	StateMachineRef   *persistencespb.StateMachineRef
+	StateMachinesTask Task
+}
+
+// StateMachinesKey gets the state machine [Store] [Key] for from this reference.
+func (t TaskWithReference) StateMachinesKey() Key {
+	return Key{Type: t.StateMachineRef.Type, ID: t.StateMachineRef.Id}
+}
+
+// MutableState is a top level container for state machines.
+// It is intentionally kept very minimal in order to support different implementations and not tie it to a specific
+// state machine's functionality. For example, a callback executor might need to extract the nexus completion from a
+// MutableState object but not all mutable states are required to support that.
+// In the future this may be merged into the [StateMachine] interface to allow state machines to have a hierarchy.
+type MutableState interface {
```

Comment:
I really like that you've broken the dependency cycle here in a way where we don't rely on the workflow mutable state, but it seems weird that we still need to be aware of the "MutableState" concept in this package, especially just to retrieve a `Store`. However, I also understand that we can't just pass in a `*Store` here because we need other mutable state methods in the callback task executor. However, we don't want to make this one big object. Could we instead make this a generic parameter on the transition function if to support different environments?

Reply:
This would require that Executor is generic which is hard to work with in Go.
I plan on iterating on this when we add the next state machine.
I definitely don't consider this final.